quicktest update & merge CerebNet.utils.metrics into FastSurferCNN.utils.metrics

taha-abdullah opened the pull request:
- Deleted CerebNet.utils.metrics.py and merged all functionality into FastSurferCNN.utils.metrics.py
- refactoring usage across files
- changed dice score implementation to numpy (tried MONAI and Ignite, didn't work out)
- added support for multiple images in test_images
- nested parameterization (subjects and test files) in test_images and test_stats

taha-abdullah committed:
Merging all functionality from CerebNet.utils.metrics.py into FastSurferCNN.utils.metrics.py
- Deleting CerebNet.utils.metrics.py
- refactoring usage across files
- Changing back to numpy dice score implementation
- removing CerebNet.utils.metrics.py from docs
- Parameterized tests in test_images.py and test_stats.py
- also added test configuration files for new filetypes
- rebased pytest-feture onto merge-metrics
- adding support for multiple images in test_images.py

taha-abdullah committed:
Rename quick_test to quicktest for more consistent spelling

dkuegler commented on requirements.txt:
Why is there a change of requirements.txt?

taha-abdullah replied:
removed extra changes. they were left in from ignite implementation

dkuegler commented on pyproject.toml:
```diff
@@ -50,6 +50,7 @@ dependencies = [
     'torchvision>=0.15.2',
     'tqdm>=4.65',
     'yacs>=0.1.8',
+    'ignite>=0.4.3',
```
maybe remove ignite again

taha-abdullah replied:
removed ignite

dkuegler commented on .github/workflows/quicktest.yaml:
Please rebase onto the current file here

taha-abdullah replied:
Done

dkuegler commented on .github/workflows/quicktest.yaml:
I do not understand why these changes are required. I think the previous version (before your changes here) was better. It has the whole pipeline grouped into one job.

I guess the "on" variable should be uncommented :)

m-reuter replied:
I commented out the on pull requests to dev and stable. These should only be added at the very end, when everything works (if at all). Until then it should be triggered manually for testing. 
Even later I think - given the computational overhead - it makes sense to trigger this only for specific pull requests, that even touch any of the files involved in testing, and not just modification to doc strings etc.

taha-abdullah replied:
Commented out trigger on pull request and changed it back to last iteration

dkuegler replied:
@m-reuter One major disadvantage for this is that such workflows can only be triggered for branches that are part of the repository. I.e. we can no longer run them on all pull requests when we want to. At least I have not found a way to tigger them. The best solution is probably to always trigger them, but have a label in the PR that needs to be set to actually run this workflow.

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
 
-        Parameters
-        ----------
+        Parameters:
+        -----------
```
```suggestion
        Parameters
        ----------
```

taha-abdullah replied:
Done

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
-        Check the output type.
+        Check the type of the output and raise an error if it doesn't match expectations.
+
+        Parameters:
+        -----------
```
```suggestion
        Parameters
        ----------
```

taha-abdullah replied:
Done

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
-        initialized and available, device is set to `cuda`.
+    Accumulating the component of the dice coefficient i.e. the union and intersection.
+
+    Args:
+        op (callable): a callable to update accumulator. Method's signature is `(accumulator, output)`.
+            For example, to compute arithmetic mean value, `op = lambda a, x: a + x`.
+        output_transform (callable, optional): a callable that is used to transform the
+            :class:`~ignite.engine.Engine`'s `process_function`'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+        device (str of torch.device, optional): device specification in case of distributed computation usage.
+            In most of the cases, it can be defined as "cuda:local_rank" or "cuda"
+            if already set `torch.cuda.set_device(local_rank)`. By default, if a distributed process group is
+            initialized and available, device is set to `cuda`.
```
Fix this docstring so it is also Numpy-style
```
Parameters
----------
ope : callable
    A callable ...
...
```

taha-abdullah replied:
Done

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
-        :class:`~ignite.engine.Engine`'s `process_function`'s output into the
-        form expected by the metric. This can be useful if, for example, you have a multi-output model and
-        you want to compute the metric with respect to one of the outputs.
-    device : str of torch.device, optional
-        Device specification in case of distributed computation usage.
-        In most of the cases, it can be defined as "cuda:local_rank" or "cuda"
-        if already set `torch.cuda.set_device(local_rank)`. By default, if a distributed process group is
-        initialized and available, device is set to `cuda`.
```
It seems to me that this metrics functionality of this class was previously implemented here and is now lost. Was this ever really used before. We should double-check. It might have been used for multi-gpu support or something.

It seems to me like FastSurfer still uses this (so at a minimum, it needs to be adapted there)
https://github.com/Deep-MI/FastSurfer/blob/462b00dc920bc5666da8c6df601afb0595737312/FastSurferCNN/utils/meters.py#L75

This device variable is being passed through the pipeline including `utils/train.py` but it only seems to be `self.device` which is initialized in init
https://github.com/Deep-MI/FastSurfer/blob/5bd139e20d8a017057550f435da6234396219206/FastSurferCNN/train.py#L79

In the end you will need to test the train script with a minimal dataset to make sure it does not get broken. I will find such a dataset for you, but if can also be generated with `generate_dataset.py`

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
+                )
+                self.union[i, j] = self.union[i, j] + torch.sum(gt) + torch.sum(pred)
+
+    def update(self, output):
```
It seems to me like update always calls `_update_union_intersection`.

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
-    def update(self, output: tuple[Any, Any], cnf_mat: bool):
+        # self.union.to(batch_output.device)
+        # self.intersection.to(batch_output.device)
+        for i, c1 in enumerate(self.class_ids):
+            gt = (labels_batch == c1).float()
+            for j, c2 in enumerate(self.class_ids):
+                pred = (batch_output == c2).float()
+                self.intersection[i, j] = self.intersection[i, j] + torch.sum(
+                    torch.mul(gt, pred)
+                )
+                self.union[i, j] = self.union[i, j] + torch.sum(gt) + torch.sum(pred)
```
Also, it seems like the original implementation had a condition here that would only compute the full confusion matrix if explicitly required. Most times, the diagonal of the confusion matrix is enough (see previous). And always computing the full confusion matrix is EXTREMELY computationally expensive. (Prohibitively so). So we should only compute the full confusion matrix if it is explicitly requested.

dkuegler commented on FastSurferCNN/utils/metrics.py:
```diff
@@ -103,175 +107,317 @@ def precision_recall(
+        if not isinstance(y, torch.Tensor):
+            y = torch.from_numpy(y)
 
-    def comput_dice_cnf(self):
```
This would be the confusion matrix :)

dkuegler reviewed (changes requested):
We should do another iteration on metrics.

taha-abdullah committed:
PR 590 Review changes:
- quicktest.yaml: commenting out trigger on pull request
- FastSurferCNN.utils.metrics.py: docstring changes

dkuegler committed:
Fix == true comparisons in quicktest

dkuegler committed:
Lots of changes to quicktest

- Add author to github actions
- Fix strings in quicktest workflow
- Add parameter options pr/**** for docker-image parameters (in addition to build-cached), which allows for building from prs via manually triggered actions
- Merge run and test of quicktest into one job to reduce clutter in reported checks
- Add cleanup option for run-fastsurfer action
- Add annotation with test results job to the quicktest workflow
- Clean up some formatting
- some minor bugfixes to FastSurferCNN.utils.common.SubjectDirectory
- Add overloads to FastSurferCNN.utils.mapper.Mapper.__call__ and some minor bugfixes to the Mapper class to make it a bit more straighforward to use
- Make the interface to FastSurferCNN.utils.mapper.TSVLookupTable by default more compatible with LUT.tsv and LUT.txt files (automatically detecting the correct format to read the file).
- Add dataframe_to_table (reverse to table_to_dataframe) and read_statsfile to FastSurferCNN.segstats (as importable functions that may be used to handling stats files and stats data)
- Update and rework the full quicktest pipeline including fixtures, required environment variables for configuration and data files with thresholds

dkuegler committed:
Add functionality to export computed values into a csv file
Move thresholds files into data (instead of thresholds) and remove thresholds folder

dkuegler committed:
Add --3T flag to quicktest for 0.8mm test case

dkuegler committed:
Fix ruff errors
